File: python/src/transtex/parsing.py
"""Public citation parsing interface."""
from __future__ import annotations

from typing import Callable

from .bibtex import reference_to_bibtex
from .parsers.apa import parse_apa_citation as _parse_apa_citation_impl
from .parsers.chicago import parse_chicago_citation as _parse_chicago_citation_impl
from .parsers.ieee import parse_ieee_citation as _parse_ieee_citation_impl
from .parsers.mla import parse_mla_citation as _parse_mla_citation_impl
from .parsers.vancouver import parse_vancouver_citation as _parse_vancouver_citation_impl
from .reference import Reference


class CitationParseError(ValueError):
    """Raised when a formatted citation string cannot be parsed."""


def parse_citation(style: str, text: str) -> Reference:
    """Parse a formatted citation string into a Reference."""
    normalized_style = style.strip().lower()
    parsers: dict[str, Callable[[str], Reference]] = {
        "apa": parse_apa_citation,
        "apa6": parse_apa_citation,
        "apa7": parse_apa_citation,
        "ieee": parse_ieee_citation,
        "chicago": parse_chicago_citation,
        "mla": parse_mla_citation,
        "vancouver": parse_vancouver_citation,
    }
    parser = parsers.get(normalized_style)
    if not parser:
        raise CitationParseError(
            "Unsupported style '{style}'. Supported: apa, apa7, ieee, chicago, mla, vancouver"
        )
    return parser(text)


def citation_to_bibtex(style: str, text: str) -> str:
    """Parse a formatted citation string and serialize it to BibTeX."""
    reference = parse_citation(style, text)
    return reference_to_bibtex(reference)


def parse_apa_citation(text: str) -> Reference:
    try:
        return _parse_apa_citation_impl(text)
    except ValueError as exc:
        raise CitationParseError(str(exc)) from exc


def parse_ieee_citation(text: str) -> Reference:
    try:
        return _parse_ieee_citation_impl(text)
    except ValueError as exc:
        raise CitationParseError(str(exc)) from exc


def parse_chicago_citation(text: str) -> Reference:
    try:
        return _parse_chicago_citation_impl(text)
    except ValueError as exc:
        raise CitationParseError(str(exc)) from exc


def parse_mla_citation(text: str) -> Reference:
    try:
        return _parse_mla_citation_impl(text)
    except ValueError as exc:
        raise CitationParseError(str(exc)) from exc


def parse_vancouver_citation(text: str) -> Reference:
    try:
        return _parse_vancouver_citation_impl(text)
    except ValueError as exc:
        raise CitationParseError(str(exc)) from exc


__all__ = [
    "CitationParseError",
    "parse_citation",
    "citation_to_bibtex",
    "parse_apa_citation",
    "parse_ieee_citation",
    "parse_chicago_citation",
    "parse_mla_citation",
    "parse_vancouver_citation",
]

File: typescript/src/parsing.ts
import { referenceToBibtex } from "./bibtex";
import { Reference } from "./reference";
import {
    parseApaCitation as parseApaCitationImpl,
    parseChicagoCitation as parseChicagoCitationImpl,
    parseIeeeCitation as parseIeeeCitationImpl,
    parseMlaCitation as parseMlaCitationImpl,
    parseVancouverCitation as parseVancouverCitationImpl,
} from "./parsers";

export class CitationParseError extends Error {}

type ParserFn = (text: string) => Reference;

const PARSERS: Record<string, ParserFn> = {
    apa: parseApaCitation,
    apa6: parseApaCitation,
    apa7: parseApaCitation,
    ieee: parseIeeeCitation,
    chicago: parseChicagoCitation,
    mla: parseMlaCitation,
    vancouver: parseVancouverCitation,
};

export function parseCitation(style: string, text: string): Reference {
    const normalized = style.trim().toLowerCase();
    const parser = PARSERS[normalized];
    if (!parser) {
        throw new CitationParseError(`Unsupported style '${style}'. Supported: ${Object.keys(PARSERS).join(", ")}`);
    }
    return parser(text);
}

export function citationToBibtex(style: string, text: string): string {
    const reference = parseCitation(style, text);
    return referenceToBibtex(reference);
}

export function parseApaCitation(text: string): Reference {
    return wrapParser(parseApaCitationImpl, text);
}

export function parseIeeeCitation(text: string): Reference {
    return wrapParser(parseIeeeCitationImpl, text);
}

export function parseChicagoCitation(text: string): Reference {
    return wrapParser(parseChicagoCitationImpl, text);
}

export function parseMlaCitation(text: string): Reference {
    return wrapParser(parseMlaCitationImpl, text);
}

export function parseVancouverCitation(text: string): Reference {
    return wrapParser(parseVancouverCitationImpl, text);
}

function wrapParser(fn: ParserFn, text: string): Reference {
    try {
        return fn(text);
    } catch (err) {
        const message = err instanceof Error ? err.message : "Failed to parse citation";
        throw new CitationParseError(message);
    }
}
